x/mongo/driver/tranquil_operation.go

```diff
@@ -136,3 +153,374 @@ func (op Operation) decodeOpQuery(wm []byte) (bsoncore.Document, error) {
+				op.Client.UpdateCommitTransactionWriteConcern()
+				op.WriteConcern = op.Client.CurrentWc
+			}
+			return nil, true
+		}
+
+		if op.Client != nil && op.Client.Committing && (retryableErr || tt.Code == 50) {
+			// If we got a retryable error or MaxTimeMSExpired error, we add UnknownTransactionCommitResult.
+			tt.Labels = append(tt.Labels, UnknownTransactionCommitResult)
+		}
+		return tt, false
+	default:
+		return nil, false
+	}
+	return nil, false
+}
+
+func (op Operation) ExecuteDirect(ctx context.Context, scratch []byte) (*wiremessage.MsgHeader, []byte, error) {
+	var header *wiremessage.MsgHeader
+	var res bsoncore.Document
+	var operationErr WriteCommandError
+
+	setup, err := op.executeValidateAndConfig(ctx, scratch)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer setup.Close()
+
+	currIndex := 0
+	for {
+		if setup.batching {
+			targetBatchSize := setup.desc.MaxDocumentSize
+			maxDocSize := setup.desc.MaxDocumentSize
+			if op.shouldEncrypt() {
+				// For client-side encryption, we want the batch to be split at 2 MiB instead of 16MiB.
+				// If there's only one document in the batch, it can be up to 16MiB, so we set target batch size to
+				// 2MiB but max document size to 16MiB. This will allow the AdvanceBatch call to create a batch
+				// with a single large document.
+				targetBatchSize = cryptMaxBsonObjectSize
+			}
+
+			err = op.Batches.AdvanceBatch(int(setup.desc.MaxBatchCount), int(targetBatchSize), int(maxDocSize))
+			if err != nil {
+				// TODO(GODRIVER-982): Should we also be returning operationErr?
+				return nil, nil, err
+			}
+		}
+
+		// convert to wire message
+		if len(scratch) > 0 {
+			scratch = scratch[:0]
+		}
+		//CREATE MESSAGE
+		wm, startedInfo, err := op.createWireMessage(ctx, scratch, *setup.desc, setup.conn)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// set extra data and send event if possible
+		startedInfo.connID = setup.conn.ID()
+		startedInfo.cmdName = op.getCommandName(startedInfo.cmd)
+		startedInfo.redacted = op.redactCommand(startedInfo.cmdName, startedInfo.cmd)
+		op.publishStartedEvent(ctx, startedInfo)
+
+		// get the moreToCome flag information before we compress
+		moreToCome := wiremessage.IsMsgMoreToCome(wm)
+
+		// compress wiremessage if allowed
+		if compressor, ok := setup.conn.(Compressor); ok && op.canCompress(startedInfo.cmdName) {
+			wm, err = compressor.CompressWireMessage(wm, nil)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+
+		finishedInfo := finishedInformation{
+			cmdName:   startedInfo.cmdName,
+			requestID: startedInfo.requestID,
+			startTime: time.Now(),
+			connID:    startedInfo.connID,
+			redacted:  startedInfo.redacted,
+		}
+		//ROUND TRIP
+		// roundtrip using either the full roundTripper or a special one for when the moreToCome
+		// flag is set
+		var roundTrip = op.roundTripDirect
+		if moreToCome {
+			roundTrip = op.moreToComeRoundTripDirect
+		}
+		header, res, err = roundTrip(ctx, setup.conn, wm)
```

Comment:
the meat of the change, round trip now just pulls the header and body bytes off the wire, no interpretation and marshalling required